design-proposals/instancetype.kubevirt.io/expand-during-submission.md

```diff
@@ -0,0 +1,301 @@
+# Overview
+
+The initial design of instance types and preferences captured a point in time
+revision of the original resource for future use by the `VirtualMachine` to
+ensure we always generate the same `VirtualMachineInstance` at runtime.
+
+This additionally allows users to easily switch between classes, sizes and
+generations of these resources at some point in the future. However this
+flexibility comes at the cost of complexity within KubeVirt itself when handling
+more advanced `VirtualMachine` lifecycle flows.
+
+This design proposal aims to set out new alternative and configurable behavior
+where these revisions are no longer captured and are instead expanded into the
+`VirtualMachine`.
+
+## Motivation
+
+The complexity of managing these point in time revisions of instance types and
+preferences has grown over time as more features and versions of the CRDs have
+landed.
+
+Some of this complexity is exposed to users and third party integrations
+such as back-up tooling or user-interfaces built on top of KubeVirt. This can
+take the form of differing entry points for certain functionality such as hot
+plug to requiring knowledge of the stored revisions to allow for a valid back up
+and eventual restore of a given `VirtualMachine` to happen.
+
+## Goals
+
+* Provide simple configurables at the cluster and `VirtualMachine` level to expand
+  instance types and preferences
+
+## Non Goals
+
+* The default behavior will not change as part of this work
+
+## User Stories
+
+* As a cluster owner I want to configure the expansion of all new
+  `VirtualMachines` referencing instance types and preferences
+* As a VirtualMachine owner I want to optionally enable or disable the expansion
+  my VirtualMachine referencing an instance type or preference
+
+## Repos
+
+* kubevirt/kubevirt
+
+## Design
+
+A new `KubeVirt` configurable will be introduced to have `VirtualMachines`
+referencing instance types or preferences expanded.
+
+Additional `VirtualMachine` configurables will also be introduced to allow a
+specific newly submitted VirtualMachine referencing instance types or
+preferences to be expanded.
```

Comment:
Can we consider alternatives where this is not available? My question stems form me not being aware of the second user story (a VM owner wanting to control this). I heard only of cluster admins.

Reply:
I can capture that in alternatives but personally I think that VM Owners should have the ability to expand or not regardless of the admins choice.

Reply:
Thanks. But if I understand correctly, the biggest issue today is the complexity for third parties. If each VM owner can choose whatever they want, the complexity remains.

I would prefer have an opinionated solution, with less knobs and options; if we see a use case for the VM owners having this freedom to work against their cluster admin, we can add per-VM knob later. Removing knobs is practically impossible.

Reply:
That complexity will remain still even if this configurable is limited to cluster admins surely, especially given that we are not changing the default here.

Making this a cluster admin only configurable will mean that the end user experience of using a VirtualMachine initially defined with instance types and preferences can change drastically between environments and there's nothing a user can do to avoid this.

By providing an additional VM owner configurable they can be explicit and know that the resulting VirtualMachine will behave as expected.

Reply:
> Users will also be unable to resize their `VirtualMachines` by making a singular
choice of instance type in the future

Why? A cluster admin that would like to interact with a third party can configure `expand` on her cluster, and that's that. If we complicate the solution and let the VM owner choose his own semantic, the cluster admin is left with uncertainty.

> Making this a cluster admin only configurable will mean that the end user experience of using a VirtualMachine initially defined with instance types and preferences can change drastically between environments and there's nothing a user can do to avoid this.

Correct. And that's exactly what I, as a cluster admin, want. I need to interact with a third party that is unaware of instance types, so I want my cluster to comply.

> By providing an additional VM owner configurable they can be explicit and know that the resulting VirtualMachine will behave as expected.

Right, but this adds complexity to the code, which I am not aware was requested by any user, increases development time, expands test matrix and add possibility of bugs, and cannot be dropped out of the API in the future.

Reply:
Okay these are fair points, I'll remove the user configurables for now and we can always follow up in the future if requested.